File: src/lib/scrapers/remoteok.ts
import type { RawPosting } from "@/types";
import { BaseScraper } from "./base";

interface RemoteOkJob {
  id?: string | number;
  url?: string;
  slug?: string;
  position?: string;
  company?: string;
  location?: string;
  description?: string;
  date?: string;
  [key: string]: unknown;
}

export class RemoteOkScraper extends BaseScraper {
  constructor(contactEmail: string) {
    super("remoteok", contactEmail);
  }

  async fetch(limit: number): Promise<RawPosting[]> {
    const origin = "https://remoteok.com";
    const allowed = await this.respectRobots(origin, "/api");
    if (!allowed) return [];

    const response = await this.fetchWithRetry(`${origin}/api`);
    const data = (await response.json()) as RemoteOkJob[];
    const jobs = data.slice(1, limit + 1);

    return jobs.map((job) => {
      const externalId = String(job.id ?? job.slug ?? "");
      const url = job.url ?? `${origin}/remote-jobs/${job.slug ?? externalId}`;
      return {
        source: "remoteok" as const,
        externalId,
        url,
        title: job.position ?? "Unknown role",
        company: job.company ?? "Unknown company",
        location: job.location ?? "Remote",
        description: job.description ?? "",
        postedAt: job.date ? new Date(job.date) : null,
        raw: job as Record<string, unknown>,
      };
    });
  }
}

File: scraper/scrapers/remoteok.py
from typing import Any

from .base import BaseScraper


class RemoteOkScraper(BaseScraper):
    def __init__(self, contact_email: str):
        super().__init__("remoteok", contact_email)

    def fetch(self, limit: int) -> list[dict[str, Any]]:
        origin = "https://remoteok.com"
        if not self._respect_robots(origin, "/api"):
            return []

        response = self._fetch_with_retry(f"{origin}/api")
        data = response.json()
        jobs = data[1 : limit + 1] if isinstance(data, list) else []

        postings: list[dict[str, Any]] = []
        for job in jobs:
            external_id = str(job.get("id") or job.get("slug") or "")
            slug = job.get("slug") or external_id
            postings.append(
                {
                    "source": "remoteok",
                    "externalId": external_id,
                    "url": job.get("url") or f"{origin}/remote-jobs/{slug}",
                    "title": job.get("position") or "Unknown role",
                    "company": job.get("company") or "Unknown company",
                    "location": job.get("location") or "Remote",
                    "description": job.get("description") or "",
                    "postedAt": job.get("date"),
                    "raw": job,
                }
            )
        return postings
